src/contract_client.rs

```diff
@@ -275,3 +342,182 @@ impl<Type> ContractClient<Type> {
         Ok(tx)
     }
 }
+
+/// A builder to simplify sending smart contract updates.
+pub struct ContractUpdateBuilder {
+    payload:    UpdateContractPayload,
+    sender:     AccountAddress,
+    energy:     Energy,
+    expiry:     Option<TransactionTime>,
+    add_energy: Option<Energy>,
+    nonce:      Option<Nonce>,
+    client:     v2::Client,
+}
+
+impl ContractUpdateBuilder {
+    /// Construct a new builder.
+    pub fn new(
+        client: v2::Client,
+        sender: AccountAddress,
+        payload: UpdateContractPayload,
+        energy: Energy,
+    ) -> Self {
+        Self {
+            payload,
+            sender,
+            energy,
+            expiry: None,
+            add_energy: None,
+            nonce: None,
+            client,
+        }
+    }
+
+    /// Add extra energy to the call.
+    pub fn extra_energy(mut self, energy: Energy) -> Self {
+        self.add_energy = Some(energy);
+        self
+    }
+
+    /// Set the expiry time for the transaction. If not set the default is one
+    /// hour from the time the transaction is signed.
+    pub fn expiry(mut self, expiry: TransactionTime) -> Self {
+        self.expiry = Some(expiry);
+        self
+    }
+
+    /// Set the nonce for the transaction. If not set the default behaviour is
+    /// to get the nonce from the connected [`Client`](v2::Client) at the
+    /// time the transaction is sent.
+    pub fn nonce(mut self, nonce: Nonce) -> Self {
+        self.nonce = Some(nonce);
+        self
+    }
+
+    /// Return the amount of [`Energy`] allowed for execution if
+    /// the transaction was sent with the current parameters.
+    pub fn current_energy(&self) -> Energy { self.energy + self.add_energy.unwrap_or(100.into()) }
```

Comment:
Maybe store the `100` in a const as there are multiple references to this number. And is there a specific reason for this behaviour? Why is this not 0 for instance (or 1, if the reason is the "off by one" thing)?

Reply:
100 is a large small number :) 

0 is too risky, things can change in between the calls. 100 is to give some buffer. But I don't have an argument for 100 vs. 90.

Ideally this would be 10% instead. I'll change it to min(50, 10%).

Reply:
It's now changed to this and only occurs in one place, this function.